docs/config/config.md
---
layout: default
title: Configuration
description: Configuring a kosher project.
has_children: true
nav_order: 20
---

# Configuration

kosher uses 4 configuration files to define your environments, pages, page elements, and test behavior.
{: .fs-6 .fw-300 }

docs/config/settings.md
---
layout: default
title: settings.json
description: Options for settings.json
parent: Configuration
nav_order: 4
---

# settings.json

The `settings.json` file allows the user to specify global settings for a kosher project. Some values in the file are defaults but can be overridden by flags of the `run` command (e.g. `--appVersion`, `--environment`).
{: .fs-6 .fw-300 }

## Format

```json
{
    "projectName": "kosher",
    "appVersion": "1.0.0",
    "driver": "chrome",
    "reportFormat": "pretty",
    "dateFormat": "MM/DD/YYYY",
    "defaultEnvironment": "test",
    "screenFormat": "desktop",
    "quitOnFail": false,
    "ignoreInvisible": true,
    "waitAfterScenario": 0,
    "waitAfterStep": 0,
    "accessibility": {
		"ruleSets": [
			"wcag21aa",
			"section508"
		],
		"impactThreshold": "serious"
	},
    "screenFormats": {
        "desktop": {
            "width": 2000,
            "height": 980
         },
        "mobile": {
            "width": 362,
            "height": 868
        },
        "tablet": {
            "width": 814,
            "height": 868
        },
        "landscape": {
            "width": 522,
            "height": 362
        }
    },
	"integrations": {
		"jira": {
			"host": "http://127.0.0.1",
			"defaults": {
				"projectKey": "PROJE",
				"issueType": "Bug",
				"affectsVersion": "1.0.0",
				"labels": "test,functional,kosher",
				"priority": "Normal"
			}
		}
	}
}
```

## Properties

### projectName

Sets the name of the project recorded in results files.

| Characteristic | Description       |
| -------------- | ----------------- |
| type           | string            |
| required       | no                |
| default        | kosher tested app |

### appVersion

Sets the version of the project recorded in results files.

| Characteristic | Description |
| -------------- | ----------- |
| type           | string      |
| required       | no          |
| default        | 1.0.0       |

### driver

Specifies which webdriver is used to execute the tests. Valid options are `chrome`, `ie`, `phantomjs`, `desktop`

| Characteristic | Description |
| -------------- | ----------- |
| type           | string      |
| required       | no          |
| default        | chrome      |

### reportFormat

Specifies what format is used to report test results. Valid options are: `pretty`, `bootstrap`, `simple`, `progress`, `junit`, `cucumber-json`

For descriptions of the reports, see [Reports](../reports/reports.html)

| Characteristic | Description          |
| -------------- | -------------------- |
| type           | string, string array |
| required       | no                   |
| default        | pretty               |

You can use multiple formats by setting the `reportFormat` to a JSON string array: `["pretty", "bootstrap"]`

### dateFormat

Specifies the format used when validating or setting dates.

| Characteristic | Description |
| -------------- | ----------- |
| type           | string      |
| required       | no          |
| default        | MM/DD/YYYY  |

Valid placeholders are:

| Placeholder | Description   | Example |
| ----------- | ------------- | ------- |
| MMMM        | Full month    | January |
| MMM         | Short month   | Jan     |
| MM          | Numeric month | 01      |
| YYYY        | Full year     | 2006    |
| YY          | Short year    | 06      |
| DDDD        | Full day      | Monday  |
| DDD         | Short day     | Mon     |
| DD          | Numeric day   | 02      |

### defaultEnvironment

Specifies the default environment to execute tests in. For more information, see [environments.json](environments.html).

| Characteristic | Description |
| -------------- | ----------- |
| type           | string      |
| required       | no          |
| default        | test        |

### screenFormat

Specifies which dimensions to apply to newly opened window and when the `I maximize the window` step is executed. For more information, see [screenFormats](#screenformats) below.

| Characteristic | Description |
| -------------- | ----------- |
| type           | string      |
| required       | no          |
| default        | desktop     |

### quitOnFail

Specifies if kosher continues executing remaining scenarios after a step fails.

| Characteristic | Description |
| -------------- | ----------- |
| type           | boolean     |
| required       | no          |
| default        | false       |

### ignoreInvisible

Specifies if kosher should act as if elements that are invisible to the user do not exist.

| Characteristic | Description |
| -------------- | ----------- |
| type           | boolean     |
| required       | no          |
| default        | true        |

### waitAfterScenario

Specifies a wait time to pause for following every scenario. **This is measured in milliseconds.**

| Characteristic | Description |
| -------------- | ----------- |
| type           | int         |
| required       | no          |
| default        | 0           |

### waitAfterStep

Specifies a wait time to pause for following every step. **This is measured in milliseconds.**

| Characteristic | Description |
| -------------- | ----------- |
| type           | int         |
| required       | no          |
| default        | 0           |

### accessibility

Configures how the Axe accessiblity API will scan the page for issues.

#### ruleSets

The Axe tags to apply to the page. These are effectively sets of rules. You can find a list of available rule sets at [Axe API Documentation](https://www.deque.com/axe/axe-for-web/documentation/api-documentation/#api-name-axegetrules).

| Characteristic | Description                           |
| -------------- | ------------------------------------- |
| type           | string array                          |
| required       | no                                    |
| default        | [ "wcag2a", "wcag2aa", "section508" ] |

#### impactThreshold

Specifies the impact (i.e. severity) that an accessibility finding must be to fail the [I test the page for accessibility](../steps/i_test_for_accessibility.html) step. See the [Axe API Documentation](https://www.deque.com/axe/axe-for-web/documentation/api-documentation/#results-object) for more details.

| Characteristic    | Description                        |
| ----------------- | ---------------------------------- |
| type              | string                             |
| required          | no                                 |
| default           | critical                           |
| available options | minor, moderate, serious, critical |


### screenFormats

Defines available screen formats. One of these must be selected by the [screenFormat](#screenformat) option above.

Each option in the `screenFormats` object must define a `width` and `height` property, both of type `int`.

<table>
    <thead>
        <tr>
            <th>Characteristic</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>type</td>
            <td>object</td>
        </tr>
        <tr>
            <td>required</td>
            <td>no</td>
        </tr>
        <tr>
            <td>default</td>
            <td><pre>
{
    "desktop": {
        "width": 2000,
        "height": 980
    },
    "mobile": {
        "width": 362,
        "height": 868
    },
    "tablet": {
        "width": 814,
        "height": 868
    },
    "landscape": {
        "width": 522,
        "height": 362
    }
}</pre>
            </td>
        </tr>
    </tbody>
</table>

config/settings.go
package config

import (
	"bytes"
	"log"
	"os"

	"github.com/cbush06/kosher/common"
	"github.com/cbush06/kosher/fs"
	"github.com/spf13/afero"
	"github.com/spf13/viper"
)

// Settings wraps references to all setting providers required by Kosher
type Settings struct {
	Environments Provider
	Pages        Provider
	Selectors    Provider
	Settings     Provider
	FileSystem   *fs.Fs
}

// NewSettings attempts to build a Settings object based on the given Fs object
func NewSettings(fs *fs.Fs) *Settings {
	settings := &Settings{
		Environments: buildProvider(common.EnvironmentsFile, fs, nil),
		Pages:        buildProvider(common.PagesFile, fs, modPagesProvider),
		Selectors:    buildProvider(common.SelectorsFile, fs, nil),
		Settings:     buildProvider(common.SettingsFile, fs, modSettingsProvider),
		FileSystem:   fs,
	}
	return settings
}

// GetEnvironmentBaseURL returns the base URL for the environment of the current run
func (s *Settings) GetEnvironmentBaseURL() string {
	if !s.Settings.IsSet("environment") {
		log.Print("No setting found for [environment]\n")
		return ""
	}

	environment := s.Settings.GetString("environment")
	if len(environment) < 1 {
		log.Print("Empty setting found for [environment]\n")
		return ""
	}

	if !s.Environments.IsSet(environment) {
		log.Printf("No entry found for [%s] in the environments file\n", environment)
		return ""
	}
	return s.Environments.GetString(environment)
}

type providerModifier func(v *viper.Viper)

func buildProvider(fileName string, fs *fs.Fs, modifyProvider providerModifier) Provider {
	var (
		provider = viper.New()
		path, _  = fs.ConfigDir.RealPath(fileName)
		exists   bool
		rawBytes []byte
		file     afero.File
		err      error
	)

	if exists, err = afero.Exists(fs.ConfigDir, fileName); !exists || err != nil {
		if err != nil {
			log.Println(err)
		}
		log.Println("Configuration file does not exist: " + path)
		return nil
	}

	if file, err = fs.ConfigDir.OpenFile(fileName, os.O_RDONLY, 0744); err != nil {
		log.Printf("Error encountered while opening config file at [%s]: %s\n", path, err)
		return nil
	}

	if rawBytes, err = afero.ReadAll(file); err != nil {
		log.Printf("Error reading config file [%s]: %s\n", path, err)
		return nil
	}

	provider.SetConfigType("json")
	provider.ReadConfig(bytes.NewBuffer(rawBytes))

	if modifyProvider != nil {
		modifyProvider(provider)
	}

	return provider
}

func modPagesProvider(v *viper.Viper) {
	v.SetDefault("home", "/")
}

func modSettingsProvider(v *viper.Viper) {
	v.SetDefault("cucumberDialect", "en")
	v.SetDefault("projectName", "kosher tested app")
	v.SetDefault("appVersion", "1.0.0")
	v.SetDefault("platform", "web")
	v.SetDefault("driver", "chrome")
	v.SetDefault("reportFormat", "pretty")
	v.SetDefault("dateFormat", "MM/DD/YYYY")
	v.SetDefault("defaultEnvironment", "test")
	v.SetDefault("screenFormat", "desktop")
	v.SetDefault("quitOnFail", false)
	v.SetDefault("ignoreInvisible", true)
	v.SetDefault("waitAfterScenario", 0)
	v.SetDefault("waitAfterStep", 0)
	v.SetDefault("accessibility.ruleSets", []string{"wcag2a", "wcag2aa", "section508"})
	v.SetDefault("accessibility.impactThreshold", "serious")
	v.SetDefault("screenFormats.desktop.width", 2000)
	v.SetDefault("screenFormats.desktop.height", 980)
	v.SetDefault("screenFormats.mobile.width", 362)
	v.SetDefault("screenFormats.mobile.height", 868)
	v.SetDefault("screenFormats.tablet.width", 814)
	v.SetDefault("screenFormats.tablet.height", 868)
	v.SetDefault("screenFormats.landscape.width", 522)
	v.SetDefault("screenFormats.landscape.height", 362)
	v.SetDefault("integrations.jira.host", "http://127.0.0.1")
	v.SetDefault("integrations.jira.defaults.projectKey", "PROJE")
	v.SetDefault("integrations.jira.defaults.issueType", "Bug")
	v.SetDefault("integrations.jira.defaults.affectsVersion", "1.0.0")
	v.SetDefault("integrations.jira.defaults.labels", "test,functional,kosher")
	v.SetDefault("integrations.jira.defaults.priority", "Normal")
}
